File: packages/api-models-ts/src/Listings.ts
/**
 * Request and response models for `/api/listing`.
 *
 * @since 1.0.0
 */

import * as v from "valibot";

import * as Schemas from "./Schemas.ts";

/**
 * An object saved by a user to a named list (upstream `Listing`).
 *
 * The handler returns bare `Listing` rows, so the `user` and `obj`
 * relationships are never loaded and are not declared here.
 *
 * @since 1.0.0
 * @category Models
 */
export const Listing = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        user_id: Schemas.NullishInteger,
        obj_id: Schemas.NullishString,
        list_name: Schemas.NullishString,
        params: Schemas.nullish(Schemas.JsonObject),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type Listing = v.InferOutput<typeof Listing>;

/**
 * Payload for adding an object to a user's list.
 *
 * `list_name` is user-defined and must start with an alphanumeric character or
 * underscore; `"favorites"` and `"rejected_candidates"` have special meaning
 * in the web app. `user_id` defaults to the token's own user, and only admins
 * may add listings to other users' accounts. `params` is required for the
 * `"watchlist"` list and must contain numeric `arcsec` (0 to 3600) and
 * `cadence` (1 or more) keys, plus an optional boolean `end_of_night`. A given
 * user, object, and list name combination may only be saved once.
 *
 * @since 1.0.0
 * @category Models
 */
export interface ListingPost {
    readonly obj_id: string;
    readonly list_name: string;
    readonly user_id?: number | undefined;
    readonly params?: Record<string, unknown> | undefined;
}

/**
 * Result of adding an object to a user's list.
 *
 * @since 1.0.0
 * @category Models
 */
export const ListingPostResponse = Schemas.model(v.strictObject({ id: Schemas.Integer }));

/**
 * @since 1.0.0
 * @category Models
 */
export type ListingPostResponse = v.InferOutput<typeof ListingPostResponse>;

File: packages/api-models-py/skyportal_py_models/listings.py
"""Request and response models for ``/api/listing``."""

from __future__ import annotations

import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class Listing(BaseModel):
    """An object saved by a user to a named list (upstream ``Listing``)."""

    # The handler returns bare ``Listing`` rows, so the ``user`` and ``obj``
    # relationships are never loaded and are not declared here.

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime.datetime | None = None
    modified: datetime.datetime | None = None
    user_id: int | None = None
    obj_id: str | None = None
    list_name: str | None = None
    params: dict[str, Any] | None = None


class ListingPost(BaseModel):
    """Payload for adding an object to a user's list."""

    model_config = ConfigDict(extra="forbid")

    obj_id: str
    list_name: str
    user_id: int | None = None
    params: dict[str, Any] | None = None


class ListingPostResponse(BaseModel):
    """Result of adding an object to a user's list."""

    model_config = ConfigDict(extra="forbid")

    id: int
